lib/shoryuken/middleware/chain.rb

```diff
@@ -93,11 +93,13 @@ def clear
 
       def invoke(*args, &final_action)
         chain = retrieve.dup
-        traverse_chain = lambda do
+        traverse_chain = lambda do |*modified_args|
+          next_args = modified_args.any? ? modified_args : args
```

Comment:
I like the overall idea here and I like especially how this is backwards-compatible with existing middleware that just calls `yield`. The implementation, however, has an interesting quirk that I'm not sure was intended: Suppose you have three middleware A, B, and C. A modifes the args, but B does not. When it's time to invoke `C`, `modified_args` is empty (B didn't yield anything), so `C` receives the original top-level `args`. Instead, it probably makes more sense for `C` to receive the args as they were passed to B (the same args that A `yield`ed).